### DISCUSSION TITLE
请问能否详细描述一下opencl和gpu的关系

### DISCUSSION THREAD

**Discussion by c4midori** [THU-DSP-LAB/ventus-gpgpu#65] [General]
我看了ventus-gpgpu-master中的代码和文档，已经了解了gpgpu的结构和怎么针对rtl进行仿真。但是我对怎么使用opencl还是不理解，不知道opencl编译出来的代码怎么执行。
能不能专门写个文档，描述一下opencl编译出来的代码是怎么在gpu上运行的？

**Comment by reoLantern** [comment]
我不是特别了解软件工具链的细节，但是可以大概描述一下。
我们的[软件工具链](https://github.com/THU-DSP-LAB/llvm-project)包含llvm编译器、pocl runtime、硬件驱动。opencl程序包含host端和device端代码（kernel）。host端代码首先执行，启动pocl runtime，对opencl api的调用通过ocl-icd链接到pocl，host端会设置kernel参数、内存声明等。pocl会调用device端的编译器编译kernel，调用ventus-driver启动spike模拟器。这些调用过程都是通过动态库的方式实现的。
spike执行kernel后，会生成.data和.metadata文件。这2个文件可以复制到chisel仓库进行RTL仿真，例如chisel仓库的ventus/txt文件夹下就是一些测试用例。
关于软件工具链的技术介绍，可以参考[在2024年RISC-V中国峰会的演讲（软件工具链部分）](https://www.bilibili.com/video/BV1KjtCeFEMC)。
欢迎继续交流！

**Reply by c4midori** [reply]
非常感谢
